ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2836,230 @@ public sealed class Bot : IAsyncDisposable {
 			}
 		}
 
+		private void OnInventoryChanged() {
+			Utilities.InBackground(CardsFarmer.OnNewItemsNotification);
+
+			if (BotConfig.BotBehaviour.HasFlag(BotConfig.EBotBehaviour.DismissInventoryNotifications)) {
+				Utilities.InBackground(ArchiWebHandler.MarkInventory);
+			}
+
+			if (BotConfig.SendSetsOnCompleted) {
+				Utilities.InBackground(SendCompletedSets);
+			}
+		}
+
+		private async Task<HashSet<uint>> GetCompletedBadgeAppIDs() {
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(1).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return new HashSet<uint>(0);
+			}
+
+			// We need a collection that does not discard duplicates, as we will get the same ID for normal as well as foil badges
+			List<uint> appIDs = GetCompletedBadgeAppIDs(badgePage);
+
+			byte maxPages = 1;
+			IElement? htmlNode = badgePage.SelectSingleNode("(//a[@class='pagelink'])[last()]");
+
+			if (htmlNode != null) {
+				string lastPage = htmlNode.TextContent;
+
+				if (string.IsNullOrEmpty(lastPage)) {
+					ArchiLogger.LogNullError(nameof(lastPage));
+
+					return new HashSet<uint>(0);
+				}
+
+				if (!byte.TryParse(lastPage, out maxPages) || (maxPages == 0)) {
+					ArchiLogger.LogNullError(nameof(maxPages));
+
+					return new HashSet<uint>(0);
+				}
+			}
+
+			const byte maxBadgesPerPage = 150;
+			byte currentPage = 2;
+
+			while ((currentPage < maxPages) && (appIDs.Count == currentPage * maxBadgesPerPage)) {
+				appIDs.AddRange(await GetCompletedBadgeAppIDs(currentPage++).ConfigureAwait(false));
+			}
+
+			return appIDs.ToHashSet();
+		}
+
+		private async Task<List<uint>> GetCompletedBadgeAppIDs(byte page) {
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(page).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return new List<uint>(0);
+			}
+
+			return GetCompletedBadgeAppIDs(badgePage);
+		}
+
+		private List<uint> GetCompletedBadgeAppIDs(IDocument badgePage) {
+			List<IElement> craftButtons = badgePage.SelectNodes("//a[@class='badge_craft_button']");
```

Comment:
This method will skip games with fully crafted badges, even if we still have any sets.

Reply:
How should we detect that? This edge case does not show at all in badges page. Checking how many cards are required for every single game we have cards of is not an option.

Reply:
Why that's not an option? It works pretty well even with hundreds of games.
We can load completed badge pages to get info if we have any sets of them. Also I think it's possible to get all the completed sets just by loading an inventory and making use of `Trading.GetInventorySets` function, thus removing parsing badges pages entirely (except for getting card amount in each set).

Reply:
> Why that's not an option?
Because it's a ton of requests we might not even need a single one of. I think we should keep load on steam servers as minimal as possible.

I already had a look at `Trading.GetInventorySets` during first implementation. It cares about rarity, which we don't.

Reply:
> I already had a look at Trading.GetInventorySets during first implementation. It cares about rarity, which we don't.

And as I already said, you should.

Reply:
I went with the route of parsing badges overview pages and only loading set count for badges we either have the craft button for or we have maximum level already crafted.

I also changed it so we only consider cards of the same rarity to be of one set as requested by @JustArchi